File: android/app/src/main/java/ai/neurophone/widget/NeurophoneAppWidget.java
// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: 2026 Jonathan D.A. Jewell
package ai.neurophone.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;

import org.json.JSONObject;

import ai.neurophone.NativeLib;
import ai.neurophone.NeurophoneActivity;
import ai.neurophone.NeurophoneService;
import ai.neurophone.R;

/**
 * Home-screen widget (sub-issue #113, one third of the widget triple).
 *
 * <p>Thin {@link AppWidgetProvider} shim. Owns no neural state of its own:
 * every render reads the live system state straight out of
 * {@link NativeLib#getState()} (a JSON string) and {@link NativeLib#isRunning()}.
 * The "Ask" action opens {@link NeurophoneActivity}; the actual query is typed
 * there, not in the widget.
 *
 * <p>Replaces the legacy Kotlin {@code NeurophoneAppWidget.kt}. The
 * SharedPreferences-backed {@code publishState(...)} path from the legacy
 * service is gone: the Rust core is the single source of truth, and the
 * widget re-reads it on every {@code ACTION_REFRESH}.
 *
 * <p><strong>{@code NeurophoneWidgetConfigureActivity} is intentionally not
 * ported</strong> &mdash; repeating the same drop decision already recorded
 * once in this repo's history ({@code git show b0de78c}, since deleted along
 * with the rest of the legacy Kotlin tree) and in
 * {@code docs/migrations/RFC-ANDROID-KOTLIN-TO-RUST.adoc} Q3. Issue #113
 * still frames this as an open owner question, so it is repeated here rather
 * than assumed silently: the widget has no configure step and works with a
 * sensible default at install (matches Q3's "DROP" resolution).
 */
public final class NeurophoneAppWidget extends AppWidgetProvider {

    public static final String ACTION_REFRESH = "ai.neurophone.widget.ACTION_REFRESH";
    public static final String ACTION_TOGGLE = "ai.neurophone.widget.ACTION_TOGGLE";
    public static final String ACTION_QUERY = "ai.neurophone.widget.ACTION_QUERY";

    @Override
    public void onUpdate(Context context, AppWidgetManager manager, int[] ids) {
        for (int id : ids) {
            render(context, manager, id);
        }
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        final String action = intent.getAction();
        if (ACTION_REFRESH.equals(action) || ACTION_TOGGLE.equals(action)) {
            if (ACTION_TOGGLE.equals(action)) {
                toggleService(context);
            }
            final AppWidgetManager mgr = AppWidgetManager.getInstance(context);
            final int[] ids = mgr.getAppWidgetIds(
                    new ComponentName(context, NeurophoneAppWidget.class));
            for (int id : ids) {
                render(context, mgr, id);
            }
        }
    }

    /**
     * Start/stop the foreground service. Running/stopped truth is re-read
     * from the core on the next render via {@link NativeLib#isRunning()} —
     * never cached locally.
     */
    private void toggleService(Context context) {
        final Intent svc = new Intent(context, NeurophoneService.class);
        if (nativeIsRunning()) {
            context.stopService(svc);
        } else {
            context.startForegroundService(svc);
        }
    }

    /**
     * Read fresh state from the Rust core and push it into the
     * {@link RemoteViews}. No SharedPreferences, no app-side cache.
     */
    private void render(Context context, AppWidgetManager mgr, int id) {
        final RemoteViews views =
                new RemoteViews(context.getPackageName(), R.layout.widget_neurophone);

        final boolean running = nativeIsRunning();
        float salience = 0f;
        String description = null;

        // NativeLib.getState() is a JSON snapshot of SystemState
        // (crates/neurophone-core). Parsed defensively: the widget must never
        // crash the launcher on a malformed/empty payload (e.g. before the
        // service has ever run).
        try {
            final String stateJson = NativeLib.getState();
            if (stateJson != null && !stateJson.isEmpty()) {
                final JSONObject state = new JSONObject(stateJson);
                salience = (float) state.optDouble("salience", 0d);
                description = state.optString("description", null);
            }
        } catch (Throwable t) {
            // Fall back to the running/stopped string below.
            description = null;
        }

        if (salience < 0f) {
            salience = 0f;
        } else if (salience > 1f) {
            salience = 1f;
        }

        views.setTextViewText(
                R.id.widget_state,
                description != null && !description.isEmpty()
                        ? description
                        : context.getString(running
                                ? R.string.widget_state_running
                                : R.string.widget_state_stopped));

        final int saliencePct = (int) (salience * 100f);
        views.setProgressBar(R.id.widget_salience, 100, saliencePct, false);
        views.setTextViewText(R.id.widget_salience_value, saliencePct + "%");

        views.setOnClickPendingIntent(R.id.widget_toggle, actionPI(context, ACTION_TOGGLE, id, 1));
        views.setOnClickPendingIntent(R.id.widget_refresh, actionPI(context, ACTION_REFRESH, id, 2));
        views.setOnClickPendingIntent(R.id.widget_query, queryPI(context, id));

        mgr.updateAppWidget(id, views);
    }

    private PendingIntent actionPI(Context context, String action, int widgetId, int requestCode) {
        final Intent intent = new Intent(context, NeurophoneAppWidget.class);
        intent.setAction(action);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId);
        final int flags = PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE;
        return PendingIntent.getBroadcast(context, requestCode * 100 + widgetId, intent, flags);
    }

    private PendingIntent queryPI(Context context, int widgetId) {
        final Intent intent = new Intent(context, NeurophoneActivity.class);
        intent.setAction(ACTION_QUERY);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        final int flags = PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE;
        return PendingIntent.getActivity(context, 1000 + widgetId, intent, flags);
    }

    /** Isolated so the one cross-language interop point is easy to audit. */
    private static boolean nativeIsRunning() {
        try {
            return NativeLib.isRunning();
        } catch (Throwable t) {
            return false;
        }
    }

    /**
     * Broadcast a refresh to every mounted widget instance. Used by
     * {@link NeurophoneWidgetActions} and other non-widget callers (service,
     * boot receiver) to nudge the widget into re-reading core state.
     */
    public static void requestRefresh(Context context) {
        final AppWidgetManager mgr = AppWidgetManager.getInstance(context);
        final int[] ids = mgr.getAppWidgetIds(new ComponentName(context, NeurophoneAppWidget.class));
        if (ids.length == 0) {
            return;
        }
        final Intent refresh = new Intent(context, NeurophoneAppWidget.class);
        refresh.setAction(ACTION_REFRESH);
        refresh.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
        context.sendBroadcast(refresh);
    }
}

File: android/app/src/main/java/ai/neurophone/widget/NeurophoneAppWidget.kt
// SPDX-License-Identifier: PMPL-1.0-or-later
// SPDX-FileCopyrightText: 2025 Jonathan D.A. Jewell
package ai.neurophone.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.widget.RemoteViews
import ai.neurophone.MainActivity
import ai.neurophone.NeurophoneService
import ai.neurophone.R

/**
 * Home-screen widget for NeuroPhone.
 *
 * Layout: title bar + state line + salience meter + "Ask" button.
 * Three intent actions:
 *   - ACTION_TOGGLE  → start/stop foreground service
 *   - ACTION_REFRESH → re-read state from prefs and redraw
 *   - ACTION_QUERY   → open MainActivity in query mode
 */
class NeurophoneAppWidget : AppWidgetProvider() {

    override fun onUpdate(
        context: Context,
        manager: AppWidgetManager,
        ids: IntArray
    ) {
        ids.forEach { id -> render(context, manager, id) }
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        when (intent.action) {
            ACTION_REFRESH, ACTION_TOGGLE, ACTION_QUERY -> {
                val mgr = AppWidgetManager.getInstance(context)
                val ids = mgr.getAppWidgetIds(ComponentName(context, NeurophoneAppWidget::class.java))
                if (intent.action == ACTION_TOGGLE) toggleService(context)
                ids.forEach { render(context, mgr, it) }
            }
        }
    }

    private fun toggleService(context: Context) {
        val prefs = prefs(context)
        val running = prefs.getBoolean(KEY_RUNNING, false)
        val intent = Intent(context, NeurophoneService::class.java)
        if (running) {
            context.stopService(intent)
            prefs.edit().putBoolean(KEY_RUNNING, false).apply()
        } else {
            context.startForegroundService(intent)
            prefs.edit().putBoolean(KEY_RUNNING, true).apply()
        }
    }

    private fun render(context: Context, mgr: AppWidgetManager, id: Int) {
        val views = RemoteViews(context.packageName, R.layout.widget_neurophone)
        val prefs = prefs(context)

        val running = prefs.getBoolean(KEY_RUNNING, false)
        val salience = prefs.getFloat(KEY_SALIENCE, 0f).coerceIn(0f, 1f)
        val description = prefs.getString(KEY_DESCRIPTION, null)

        views.setTextViewText(
            R.id.widget_state,
            description ?: context.getString(
                if (running) R.string.widget_state_running else R.string.widget_state_stopped
            )
        )
        val saliencePct = (salience * 100f).toInt()
        views.setProgressBar(R.id.widget_salience, 100, saliencePct, false)
        views.setTextViewText(R.id.widget_salience_value, "$saliencePct%")

        // Toggle (start/stop service)
        views.setOnClickPendingIntent(
            R.id.widget_toggle,
            actionPI(context, ACTION_TOGGLE, id, requestCode = 1)
        )
        // Refresh (re-render)
        views.setOnClickPendingIntent(
            R.id.widget_refresh,
            actionPI(context, ACTION_REFRESH, id, requestCode = 2)
        )
        // Ask -> open MainActivity in query mode
        views.setOnClickPendingIntent(
            R.id.widget_query,
            queryPI(context, id)
        )

        mgr.updateAppWidget(id, views)
    }

    private fun actionPI(context: Context, action: String, widgetId: Int, requestCode: Int): PendingIntent {
        val intent = Intent(context, NeurophoneAppWidget::class.java).apply {
            this.action = action
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
        }
        val flags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        return PendingIntent.getBroadcast(context, requestCode * 100 + widgetId, intent, flags)
    }

    private fun queryPI(context: Context, widgetId: Int): PendingIntent {
        val intent = Intent(context, MainActivity::class.java).apply {
            action = ACTION_QUERY
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
        }
        val flags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        return PendingIntent.getActivity(context, 1000 + widgetId, intent, flags)
    }

    companion object {
        const val ACTION_REFRESH = "ai.neurophone.widget.ACTION_REFRESH"
        const val ACTION_TOGGLE  = "ai.neurophone.widget.ACTION_TOGGLE"
        const val ACTION_QUERY   = "ai.neurophone.widget.ACTION_QUERY"

        private const val PREFS = "neurophone_widget"
        const val KEY_RUNNING = "running"
        const val KEY_SALIENCE = "salience"
        const val KEY_DESCRIPTION = "description"

        fun prefs(context: Context): SharedPreferences =
            context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

        /** Push fresh neural state from the service into the widget. */
        fun publishState(
            context: Context,
            running: Boolean,
            salience: Float,
            description: String?
        ) {
            prefs(context).edit()
                .putBoolean(KEY_RUNNING, running)
                .putFloat(KEY_SALIENCE, salience)
                .putString(KEY_DESCRIPTION, description)
                .apply()
            val mgr = AppWidgetManager.getInstance(context)
            val ids = mgr.getAppWidgetIds(ComponentName(context, NeurophoneAppWidget::class.java))
            if (ids.isNotEmpty()) {
                val intent = Intent(context, NeurophoneAppWidget::class.java).apply {
                    action = ACTION_REFRESH
                    putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
                }
                context.sendBroadcast(intent)
            }
        }
    }
}
